doc/rzil.md

```diff
@@ -0,0 +1,132 @@
+of ESIL, in particular its highly ambiguous and weak typing, requiring [major
+hacks](https://github.com/rizinorg/rizin/blob/2e065789a70edd20909aadcdf7f9c45b9af699fb/librz/analysis/esil/esil.c#L1025-L1032)
+to determine the bit-width of a value required for further calculations, for
+example.
+
+At this point, the two options to proceed were to either enhance ESIL and
+redesign some aspects for it, such as introducing static typing, or starting
+from scratch with a new language, preferably reusing an already existing
+field-tested one. It was decided for the latter approach, the language was
+chosen to be BAP Core Theory. The initial implementation was then carried out as
+part of the Rizin Summer of Code 2021 by heersin.
+
+Design
+------
+
+An expression, or "op", in RzIL forms a tree, optionally composed from multiple
+sub-ops. All these ops are statically typed and are divided into pure ops and
+effects at the highest level.
+
+### Pure Ops
+
+As the name suggests, pure ops, stored by `RzILOpPure`, are computiations that
+are free from side-effects and evaluate to a value. The types of values
+supported at the moment are booleans and bitvectors.
+
+### Effect Ops and State
+
+As the counterpart to pure ops, effects stored by `RzILOpEffect` express
+transitions from one virtual machine state to another. The state of the RzIL vm
+consists of variables, binding values to names, and memories, which are arrays
+of bitvectors indexed by bitvectors. An example of an effect op is `set n x`,
+which sets the value of the variable `n` to the value that the pure expression
+`x` evaluates to. The `seq` op allows composing a sequence of multiple effects
+to be executed after each other, `branch` introduces conditional execution and
+`repeat` enables looping.
+
+Deviations from Core Theory
+---------------------------
+
+Core Theory is very tightly integrated into the BAP ecosystem and while
+theoretically being reusable outside as-is, has certain properties that do not
+translate well to our C implementation, which is where some deliberate
+deviations were made.
+
+### Typing
+
+BAP makes heavy use of OCaml's strong type system to ensure well-typedness also
+on the IL level. In particular, any Core Theory expression that is well-typed in
+OCaml is automatically well-typed on the level of the IL.
+
+As a specific example, BAP uses the OCaml type `'s bitv` to express pure
+bitvector ops, where the size of the bitvector is statically given by the type
+variable `'s`. This ensures for example that the op `add : 's bitv -> 's bitv ->
+'s bitv` will only be able to take bitvectors of identical size as operands and
+return a bitvector of the same size as well. At the same time, it is possible to
+have polymorphic ops such as the if-then-else `ite : bool -> 'a pure -> 'a pure
+-> 'a pure` which can operate on both `bool` and `'s bitv` values (`bool` is an
+alias for `Bool.t pure` and `'s bitv` is an alias for `'s Bitv.t pure`).
```

Comment:
i'm not surehow can these be clear even for those who are not familiar with bap/ocaml stuff. i would simplify the stmts to something like `ite(bool condition, pure true_val, pure false_val)`

Reply:
Dropping the type variables makes it lose its meaning here. In fact, the entire chapter's core is explaining the differences between the exact details of BAP's OCaml code and RzIL. I've just added a sentence above to indicate that some ML knowledge is required here.